### PR TITLE
Advanced uses of decision graph

### PR DESCRIPTION
**Pull request by tomato42**
### Description

Describe how decision graph works in detail, with specific examples.

### Motivation and Context

Provide good documentation for the project

### Checklist

- [x] I have read the [CONTRIBUTING.md](https://github.com/tomato42/tlsfuzzer/blob/master/CONTRIBUTING.md) document and my PR follows [change requirements](https://github.com/tomato42/tlsfuzzer/blob/master/CONTRIBUTING.md#change-requirements) therein
- [x] the changes are also reflected in documentation and code comments
- [x] all new and existing tests pass (see Travis CI results)
- [x] [test script checklist](https://github.com/tomato42/tlsfuzzer/wiki/Test-script-checklist) was followed for new scripts
- [x] new test script added to `tlslite-ng.json` and `tlslite-ng-random-subset.json`
- [x] new and modified scripts were ran against popular TLS implementations:
  - n/a
- [x] required version of tlslite-ng updated in requirements.txt and README.md

---
This change is [Reviewable](https://reviewable.io/reviews/tomato42/tlsfuzzer/641)

### CODE REVIEW

File: docs/source/advanced-decision-tree.rst [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+any other messages, require the preceding code after
+:py:class:`~tlsfuzzer.expect.ExpectFinished`.
+That handles OpenSSL-using servers and others that behave similarly.
+
+.. note::
+    :term:`TLS` standard does allow sending NewSessionTicket messages at
+    arbitrary times after Finished.
+
+If you want the runner to finish the loop on receipt of ApplicationData
+message write:
```
Comment (The-Mule) [inline docs/source/advanced-decision-tree.rst]: 'write' sounds odd, what about: "in the following example the runner finish the loop once ApplicationData message is received."

File: docs/source/advanced-decision-tree.rst [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    join = ClientKeyExchangeGenerator()
+    # join regular path:
+    fork.next_sibling.add_child(join)
+    # join CR path:
+    node = node.add_child(join)
+
+After that, handshake continues as usual with ChangeCipherSpec, Finished, etc.
+
+.. note::
+    When specifying alternative messages, you must take care to not allow
```
Comment (The-Mule) [inline docs/source/advanced-decision-tree.rst]: take care not to allow

File: scripts/test-fuzzed-plaintext.py [hunk 3]
```diff
@@ -137,17 +141,21 @@ def add_handshake_conversation(conversations, host, port, cipher, dhe, data):
...
         node = node.add_child(ExpectServerKeyExchange())
+
+    # handle servers which ask for client certificates
     node = node.add_child(ExpectCertificateRequest())
     fork = node
     node = node.add_child(ExpectServerHelloDone())
     node = node.add_child(CertificateGenerator())
 
-    # handle servers which ask for client certificates
+    # handle servers which don't ask for client certificates
```
Comment (The-Mule) [inline scripts/test-fuzzed-plaintext.py]: that don't ask

File: scripts/test-fuzzed-plaintext.py [hunk 4]
```diff
@@ -271,17 +279,21 @@ def main():
     node = node.add_child(ExpectCertificate())
     if dhe:
         node = node.add_child(ExpectServerKeyExchange())
+
+    # handle servers which ask for client certificates
```
Comment (The-Mule) [inline scripts/test-fuzzed-plaintext.py]: that ask

File: docs/source/advanced-decision-tree.rst [hunk 5]
```diff
@@ -0,0 +1,168 @@
+.. _Decision tree:
+
+=============
+Decision tree
+=============
+
+While this documentation calls the structure traversed by the runner a
+"decision tree," it's more precisely described as a directed graphs. The name
```
[the remark this answers is not in the document]
Comment (t184256) [reply docs/source/advanced-decision-tree.rst]: Directed or oriented?

[the remark this answers is not in the document]
Comment (tomato42) [reply docs/source/advanced-decision-tree.rst]: @t184256 "A directed graph is called an oriented graph if none of its pairs of vertices is linked by two symmetric edges."
That's not forbidden, 1-cycles and 2-cycles are allowed in runner so they aren't oriented graphs

[the remark this answers is not in the document]
Comment (tomato42) [reply docs/source/advanced-decision-tree.rst]: @The-Mule yeah, it's a mess

changed the new documentation to use the "decision graph" instead "decision tree"

File: docs/source/advanced-decision-tree.rst [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+
+When a node matches received message and processes it without errors,
+runner continues execution by switching to the child.
+
+If ``child`` points to ``None``, runner closes open connections and ends
+execution.
+
+To create loops the ``child`` can point to nodes higher in the tree, including
+to itself.
+You need to use this mechanism to allow receiving arbitrary number of messages.
```
[the remark this answers is not in the document]
Comment (tomato42) [reply docs/source/advanced-decision-tree.rst]: that's passive voice, it shouldn't be used

what's wrong with it as it is now?

File: docs/source/advanced-decision-tree.rst [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+Sibling nodes
+--------------
+
+The runner uses nodes pointed to by ``next_sibling`` when received message
```
[the remark this answers is not in the document]
Comment (tomato42) [reply docs/source/advanced-decision-tree.rst]: next_sibling points to single node, but it inspects all nodes linked like that (if "→" is used for "next_sibling" and "↓" is used for "child"), so for example:
```
root
↓
A → B → C
↓   ↓   ↓
G   H   C
```
when processing node A, runner will inspect also B and C if A doesn't match the received message

File: docs/source/advanced-decision-tree.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (The-Mule) [reply docs/source/advanced-decision-tree.rst]: Nothing, really. It just sounds unintentionally like the only usage is to allow receiving arbitrary number of messages. But it is perhaps just my subjective reading of it. If you want to keep the original version, I am okay with that.

File: docs/source/advanced-decision-tree.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (The-Mule) [reply docs/source/advanced-decision-tree.rst]: Thanks for explanation.

File: docs/source/advanced-decision-graph.rst [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+==============
+
+While this documentation calls the structure traversed by the runner a
+"decision graph," as it can contain loops, it's more precisely described as a
+directed graph. Older parts of this documentation and object names refer to
+this structure as a "decision tree"—this name
+reflects the most common use case in the bundled tests, not the most
+complex supported one.
+
+Tree node fields
```
Comment (The-Mule) [inline docs/source/advanced-decision-graph.rst]: "Graph node fields" or maybe even better just "Nodes"

**Pull request merged**